### ISSUE TITLE
Segments and metrics with custom SQL

### ISSUE THREAD

**Issue by mokra** [metabase/metabase#5194]
Its a feature request.

You think it's possible?

**Comment by mazameli** [comment]
Yes, this is definitely something we're actively considering; particularly metrics based on SQL queries. 

As far as SQL-based segments, [this pull request](https://github.com/metabase/metabase/pull/4613) doesn't implement SQL segments that you can use as _filters,_ but will allow you to use the results of any saved question as the starting point for a new question.

**Comment by mokra** [comment]
I see - thats great. 

Is their any related issue? I saw a couple of issues about "SQL-Snippets" - but this is more specific I suppose

**Comment by demorenoc** [comment]
I was about to file an issue for this. Currently in the Metrics one can do with custom expressions stuff like:

`Average("Variable")`

One comon "trick" to get a proportion is to make the average of a boolean variable. In my team I usually do a lot of  SQL with expressions like ```AVG(variable > 10)``` where 10 could be any other number. With this expression I get the proportion of elements greater than 10. But in the Custom Expressions menu I cannot do the `variable > 10` piece of the aggregation I want. It would be great if at least this basic type of sub-expressions can be input in the agreggation of custom expressions.

**Comment by mokra** [comment]
@mazameli 
Since Metrics can manipulate all SQL-Query parts but `FROM`, I would suggest to divide them by this.
![89ys1zsz](https://cloud.githubusercontent.com/assets/1849512/26548536/f96aee10-4474-11e7-9ff9-5864197013e4.jpg)
The challenge will be to store it the way, that it can be used with other metrics too. How would you do it? I mean lets say we have this Metric:
```
SELECT Order.id, Order.date, order_item.name
```
`----FROM Order` --- non editable
```
LEFT JOIN Item as order_item on order_item.order_id = Order.id
WHERE order_item.id > 20
AND ....
GROUP BY order_item.name
```

But here is the question: How would one be able to use it with other Metrics and Filters? Doing regexp for `SELECT`,`WHERE`,`GROUP BY` is not a very good practice or would you have an other idea?

The other thing, I could think of, is create individual input filed for all these Parts - so for `SELECT`,`JOIN`,`WHERE`,`GROUP BY` like:
![89ys1zsz](https://cloud.githubusercontent.com/assets/1849512/26548992/10f154aa-4477-11e7-8c73-7f389b6b49de.jpg)


What do you say?

**Comment by devilankur18** [comment]
I am also trying to do something as simple as

**Comment by mokra** [comment]
I think this is not necessary any more, since nested queries exist

**Comment by jpulec** [comment]
I think this is still a feature worth having. Providing a sql based segment to allow filtering feels like a much better experience for your users, instead of requiring them to use a nested set of queries.

**Comment by tongnk** [comment]
Would really love this feature. Save so much time in defining a key set of metrics we can use throughout queries...

**Comment by lkraav** [comment]
~18 months later, is this maybe a solved problem at 0.31+ or not yet?

**Comment by tobsch** [comment]
+1

**Comment by dshadab** [comment]
+1

**Comment by develar** [comment]
It is very important feature for analytical database, for example, clickhouse gives you a lot of aggregating functions like `quantileTDigest`. But I cannot use it for metrics :(

**Comment by mazameli** [comment]
Related to SQL snippets: https://github.com/metabase/metabase/issues/1825

**Comment by mazameli** [comment]
People who are interested in this feature might to read up on the SQL snippets feature that we shipped last summer: https://www.metabase.com/learn/building-analytics/sql-templates/sql-snippets

**Comment by agnjkafgh** [comment]
I need to restrict my Field Filter to a subset of values in my Dashboard which consists of only Native Questions.
Since segments aren't working for me, is there any other way I can do something like this?

**Comment by flamber** [comment]
@agnjkafgh Try Linked Filters: https://www.metabase.com/learn/dashboards/linking-filters
Please use the forum for questions and troubleshooting: https://discourse.metabase.com/
